### ISSUE TITLE
Incorrect model returned by LightGBM calibration for multiclass with softmax

### ISSUE THREAD

**Issue by mjmckp** [dotnet/machinelearning#1424]
https://github.com/dotnet/machinelearning/blob/5123aeedb91a435236e8b94ba0f17a6d15a20db5/src/Microsoft.ML.LightGBM/LightGbmMulticlassTrainer.cs#L104

The model returned here is only correct if the objective type is `multiclassova`, since the `OvaPredictor` predictor type created here always just normalises the `_numClass` probabilities to sum to one:
https://github.com/dotnet/machinelearning/blob/5123aeedb91a435236e8b94ba0f17a6d15a20db5/src/Microsoft.ML.StandardLearners/Standard/MultiClass/Ova.cs#L575

However, when the objective type is `multiclass` (i.e., softmax), what should happen is that the `OvaPredictor` type should be given an array of predictors that return the _raw_ scores from each of the trees (i.e., the result of `CreateBinaryPredictor` instead of the `FeatureWeightsCalibratedPredictor`), and instead of normalising the outputs of these predictors, it should be computing the softmax function of the raw scores, as is done in the LightGBM implementation:
https://github.com/Microsoft/LightGBM/blob/087d30623aef4b50592cf528eb50c1c55baeaab8/src/objective/multiclass_objective.hpp#L115

These kind of bugs would be easily detected if there were unit tests comparing the results of evaluating the managed ML.NET GBDTs against the original native implementation (i.e., calling `LGBM_BoosterPredictForMat` on the Booster object used to calibrate the model).

**Comment by Zruty0** [comment]
So there's 2 issues here. 
1. Write a couple tests that ensure that ML.NET prediction output matches native LightGBM. These will now fail because of the outlined problem.
2. Fix the code so that the tests pass.

**Comment by wschin** [comment]
@mjmckp, I don't think ML.NET lightGBM supports those regression formulations because we have
```csharp
        private protected override void CheckAndUpdateParametersBeforeTraining(IChannel ch, RoleMappedData data, float[] labels, int[] groups)
        {
            Options["objective"] = "regression";
            // Add default metric.
            if (!Options.ContainsKey("metric"))
                Options["metric"] = "l2";
        }
```

**Comment by mjmckp** [comment]
It should be straightforward to support those kinds of regressions as well, right?

Btw, two other things:
* The `Str2DoubleArray` of the `Booster` object cannot parse inf and nan values generated by LightGBM.  The modified method below will, though:
```
        private static double[] Str2DoubleArray(string str, char delimiter)
        {
            return str.Split(delimiter)
                      .Select(s => double.TryParse(s.Replace("inf", "∞"), out double rslt) ? rslt : 
                                    (s.Contains("nan") ? double.NaN : throw new Exception($"Cannot parse as double: {s}")))
                      .ToArray();
        }
```
* A lot of the P/Invoke calls could be made a lot more efficient by pinning arrays - currently I think the arrays will be copied both before and after the native methods are invoked.

**Comment by wschin** [comment]
I don't know how simple it will be but I assume that we will need to implement one transform for each of the output function. Also, those formulations are used much less than squared loss.

**Comment by mjmckp** [comment]
As I mentioned above, for the Poisson, Gamma, and Tweedie regressions, the raw output of the tree needs to be made positive with a call to `Math.Exp`.

The actual values printed out by LightGBM for NaN and Inf are weird (I forget exactly, but I think they are +nan.0, +inf.0, and -inf.0), but the parsing logic above will work.

**Comment by justinormont** [comment]
I'm not convinced we should always match the output of native LightGBM. 

I believe we have a strong ability to benchmark ML.NET's version of LightGBM and optimize its hyperparameters. This should be a two-way street between ML.NET & LightGBM.

Some changes to LightGBM might entail changes to the algo/hyperparameters (which should cause some benchmarking for accuracy changes to occur on our side).

Ref: https://github.com/dotnet/machinelearning/pull/667#pullrequestreview-144937317

**Comment by wschin** [comment]
ML.NET LightGBM provides an interface to LightGBM. User may expect that this LightGBM is the same to that LightGBM.

**Comment by wschin** [comment]
Just opened another issue for other regression losses.

**Comment by mjmckp** [comment]
Thanks @wschin. What about the issue I mentioned with the `GetDefalutValue` method? What about the P/Invoke inefficiencies?

**Comment by wschin** [comment]
I don't concern numerical problems and memory usage too much for now, but you can definitely create other issues as this thread was mainly for class probability but got diverged.
